### PR TITLE
[HotFix] Stop tracking syncing live WAL for performance

### PR DESCRIPTION
**Pull request by riversand963**
With #10087, applications calling `SyncWAL()` or writing with `WriteOptions::sync=true` can suffer
from performance regression. This PR reverts to original behavior of tracking the syncing of closed WALs.
After we revert back to old behavior, recovery, whether kPointInTime or kAbsoluteConsistency, may fail to
detect corruption in synced WALs if the corruption is in the live WAL.

Test plan
make check

Before #10087
```bash
fillsync     :     750.269 micros/op 1332 ops/sec 75.027 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync     :     776.492 micros/op 1287 ops/sec 77.649 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync [AVG 2 runs] : 1310 (± 44) ops/sec;    0.1 (± 0.0) MB/sec
fillsync     :     805.625 micros/op 1241 ops/sec 80.563 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync [AVG 3 runs] : 1287 (± 51) ops/sec;    0.1 (± 0.0) MB/sec
fillsync [AVG    3 runs] : 1287 (± 51) ops/sec;    0.1 (± 0.0) MB/sec
fillsync [MEDIAN 3 runs] : 1287 ops/sec;    0.1 MB/sec
```

Before this PR and after #10087 
```bash
fillsync     :    1479.601 micros/op 675 ops/sec 147.960 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync     :    1626.080 micros/op 614 ops/sec 162.608 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync [AVG 2 runs] : 645 (± 59) ops/sec;    0.1 (± 0.0) MB/sec
fillsync     :    1588.402 micros/op 629 ops/sec 158.840 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync [AVG 3 runs] : 640 (± 35) ops/sec;    0.1 (± 0.0) MB/sec
fillsync [AVG    3 runs] : 640 (± 35) ops/sec;    0.1 (± 0.0) MB/sec
fillsync [MEDIAN 3 runs] : 629 ops/sec;    0.1 MB/sec
```

After this PR
```bash
fillsync     :     749.621 micros/op 1334 ops/sec 74.962 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync     :     865.577 micros/op 1155 ops/sec 86.558 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync [AVG 2 runs] : 1244 (± 175) ops/sec;    0.1 (± 0.0) MB/sec
fillsync     :     845.837 micros/op 1182 ops/sec 84.584 seconds 100000 operations;    0.1 MB/s (100 ops)
fillsync [AVG 3 runs] : 1223 (± 109) ops/sec;    0.1 (± 0.0) MB/sec
fillsync [AVG    3 runs] : 1223 (± 109) ops/sec;    0.1 (± 0.0) MB/sec
fillsync [MEDIAN 3 runs] : 1182 ops/sec;    0.1 MB/sec
```

### CODE REVIEW

**Review by ajkr — COMMENTED** [review]
> After we revert back to old behavior, recovery, whether kPointInTime or kAbsoluteConsistency, may fail to
detect corruption in synced WALs if the corruption is in the live WAL.

What changed after #10087 was written that led us to change our mind and reintroduce the gap in coverage? These are the events I've seen that didn't lead me to change my mind:

- MyRocks team had a test that relied on WAL with truncated synced data being recoverable. They should change the test.
- MyRocks team said the impact to them is MANIFEST becomes larger. They can reduce the manifest rollover size if it's a problem.
- MyRocks team said outside users who use rocksdb_flush_log_at_trx_commit=1 could be impacted severely. I agree with this but also knew since the feature was introduced that duplicating synced filesystem state in MANIFEST would cause per-sync overhead. Probably it should not be used in cases like rocksdb_flush_log_at_trx_commit=1 where the overhead is unaffordable.

**Comment by riversand963** [comment]
Thanks a lot for the insightful feedback @ajkr!

I should have mentioned that this PR is a hot-fix that restores to the original behavior. The purpose is to unblock MySQL/MyRocks.
In addition to this PR (after it's merged), I plan to extend track_and_verify_wals_in_manifest to be more flexible. Applications can choose whether to track sync operations on the live WAL or not. The flexibility is helpful because some applications need RocksDB to fail recovery when detecting loss of synced data in the WAL, while other applications allow RocksDB to succeed in recovery with loss of synced data, as long as the application knows which data is lost. MyRocks falls into the latter.

> MyRocks team said the impact to them is MANIFEST becomes larger. They can reduce the manifest rollover size if it's a problem.

There is more to this. Having MANIFEST files rotate faster still writes more data to flash, which is undesirable.

With the extension to `track_and_verify_wals`, applications using `rocksdb_flush_log_at_trx_commit ` can benefit from this feature, too, as long as they can use replication to tolerate synced data loss, as in the case of MyRocks.

**Review by ajkr — APPROVED** [review]
I am not used to thinking about tolerating synced data loss. I think usually the user tells us to sync something because they need it recoverable. Perhaps there are cases like `rocksdb_flush_log_at_trx_commit=1` (sync once per second), where the sync might just be a proxy for bounding replication recovery time or something. With `rocksdb_flush_log_at_trx_commit=2` (sync every write) it's hard to imagine synced data loss is tolerable.

Assuming synced data loss is tolerable, this is indeed a reasonable behavior since it still prevents holes in recovery due to clean truncations. It could also be useful in the case where synced data loss is bad but a hole in the recovered synced data is worse, and it's only affordable to protect the latter.

This seems like a narrow couple scenarios. I think the API and description do not mention it. Alternatively, it could mention that only closed WALs are protected right now, and the reader can imagine what scenarios that is useful in. I may have missed some.

**Comment by riversand963** [comment]
@ajkr 
Agree. It is case by case, and we should make it explicit in the API comment. Will update later.

**Comment by ajkr** [comment]
I think for now it can be as simple as saying only closed WALs are tracked. That is a big limitation that is not obvious in the option name or API doc.

**Comment by riversand963** [comment]
Thanks @ajkr for the review!

**Comment by riversand963** [comment]
hmm, since #10087 is already in 7.4 release, I should update HISTORY.md to mention this behavior change as well.

**Pull request closed without merging**
